### DISCUSSION TITLE
Coral TPU benchmarks

### DISCUSSION THREAD

**Discussion by Curid** [ultralytics/yolov5#6229] [Q&A]
How does YOLOv5 compare to the Tensorflow models? I'm currently using [SSDLite_MobileDet](https://www.coral.ai/models/object-detection) at 32.9 mAP and 9ms latency. 

Does anyone know of any other Coral compatible person detection models besides my [TF-CCTV](https://gitlab.com/Curid/TF-CCTV)?

**Comment by glenn-jocher** [comment]
@Curid I'm working on offering official benchmarks across most of our export pipelines here, but it's an enormous task. On the Edge TPU side, how do you benchmark? Which hardware exactly would be the 'reference' hardware, since Coral offers so many things?

<img width="1238" alt="Screen Shot 2022-01-06 at 10 32 11 AM" src="https://user-images.githubusercontent.com/26833433/148433071-a76c0db8-58d5-4e99-9dfc-26fd3736f6a3.png">

Also since these are merely accelerators I'm assuming the base system (CPU, RAM, storage) significantly influence results no? 

Do we need a Linux system for the USB accelerator or can we plug it into a Mac also?

**Comment by glenn-jocher** [comment]
To add some extra context, I want to produce official benchmarks for all our officially supported formats across any backend applicable, i.e. RPi4, Intel CPU, AMD CPU, M1 CPU, A15 ASIC (iOS), Snapdragon 888 (Android), V100 (GPU), Jetson (GPU), Coral (USB accelerator??)

## Formats

YOLOv5 inference is officially supported in 11 formats:

Format                      | `export.py --include`         | Model
:---                        | --:                           | :--
[PyTorch](https://pytorch.org/)                     | -                             | `yolov5s.pt`
[TorchScript](https://pytorch.org/docs/stable/jit.html)                 | `torchscript`                 | `yolov5s.torchscript`
[ONNX](https://onnx.ai/)                        | `onnx`                        | `yolov5s.onnx`
[OpenVINO](https://docs.openvino.ai/latest/index.html)                    | `openvino`                    | `yolov5s_openvino_model/`
[TensorRT](https://developer.nvidia.com/tensorrt)                    | `engine`                      | `yolov5s.engine`
[CoreML](https://github.com/apple/coremltools)                      | `coreml`                      | `yolov5s.mlmodel`
[TensorFlow SavedModel](https://www.tensorflow.org/guide/saved_model)       | `saved_model`                 | `yolov5s_saved_model/`
[TensorFlow GraphDef](https://www.tensorflow.org/api_docs/python/tf/Graph)         | `pb`                          | `yolov5s.pb`
[TensorFlow Lite](https://www.tensorflow.org/lite)             | `tflite`                      | `yolov5s.tflite`
[TensorFlow Edge TPU](https://coral.ai/docs/edgetpu/models-intro/)         | `edgetpu`                     | `yolov5s_edgetpu.tflite`
[TensorFlow.js](https://www.tensorflow.org/js)               | `tfjs`                        | `yolov5s_web_model/`

**Comment by Curid** [comment]
>how do you benchmark? 

I don't know what script they're using for [benchmarking.](https://www.coral.ai/docs/edgetpu/benchmarks)  I got the results from [here.](https://www.coral.ai/models/object-detection)

> Which hardware exactly would be the 'reference' hardware, since Coral offers so many things?

They all use the same chip from what I understand.

> Also since these are merely accelerators I'm assuming the base system (CPU, RAM, storage) significantly influence results no?

I don't think so, the usb accelerator can be used with a [rpi zero](https://linuxgizmos.com/google-kit-uses-rpi-zero-and-coral-accelerator-for-machine-learning/).

> Do we need a Linux system for the USB accelerator or can we plug it into a Mac also?

https://dev.to/kojikanao/use-coral-edgetpu-usb-accelerator-on-macos-435

**Reply by glenn-jocher** [reply]
@Curid got it, thanks!

> I don't think so, the usb accelerator can be used with a rpi zero.

Yes, this is what I mean. If you benchmark on a RPi vs a Linux desktop I suspect one will benchmark faster than the other even if they use the same accelerator.

**Reply by Curid** [reply]
@glenn-jocher Tensorflow uses a special coco2017 split for evaluations. Do you think that makes a significant difference to the mAP compared to your results?

https://github.com/tensorflow/models/blob/master/research/object_detection/data/mscoco_minival_ids.txt

**Reply by glenn-jocher** [reply]
@Curid yes that will certainly affect the mAP. Not sure where they got that from. The official splits are here:
https://cocodataset.org/#download

And we follow them exactly here:
https://github.com/ultralytics/yolov5/blob/33a67b4918aa3f2e572f115781e615a91fd543a4/data/coco.yaml#L10-L15

**Comment by Curid** [comment]


**Reply by glenn-jocher** [reply]
But their list is 8000 images. The official val set is 5000 images, so official val will be faster.
